Easter cleanup additions

philipppahl opened the pull request:
## 

**Changes proposed in this pull request:**
- Introduction of base class JSONSerializable for specs

**Todos**
- [x] Unit Tests
- [x] Integration Tests
- [x] Examples

/cc @amatthies

Commit:
Update PULL_REQUEST_TEMPLATE.md

philipppahl committed:
Merge remote-tracking branch 'upstream/easter_cleanup' into easter_cleanup_additions

philipppahl committed:
fixed typo in name

philipppahl committed:
first commit with new json serialization schema

philipppahl committed:
removed obsolete lines from instant_retry

amatthies commented:
-1 for making `JSONSerializable` the base class of all floto spec classes, imho that's too much inheritance just for code reuse. _Favor object composition over class inheritance_

philipppahl committed:
functional approach to serialization

philipppahl commented:
@amatthies 
Static functions are now in floto.specs.serializer and used by specs. Removed JSONSerializable.

amatthies commented:
@philipppahl Much better. Other than that:
- +1 for  removing `_deserialized` from `DeciderSpec`. Please do the same for the others.
- -1 for the method name `copy_args_wo_type`.

philipppahl commented:
@amatthies 
My idea was to have a generic deserialized in the base classes and "implementations" (_deserialized) in the child classes. I'll remove it, so that it's consistent. We then have to check the type elsewhere in the code at the deserialization step.

philipppahl committed:
JSON cleanup; improved separation of concerns in ExecutionGraph.

philipppahl commented:
@amatthies 
- removed _deserialized
- ActivityTask and specs.task.ChildWorkflow have id_ parameter for consistency
- requires parameter is now list of ids in favor of list of tasks: Better solution as we do not have to serialize the complete tasks for the requires parameter and it contains all information which is needed.

ExecutionGraph:
- Only handles the requirements, does not use tasks any more (better separation of concerns, more transparent). Cleaner solution wrt. list of ids in `requires` of tasks
- Removed transitive reduction O(n^2.4). Can become computing intense in case of many generators and many tasks. Was introduced to improve fewer queries to the history ("isTaskCompleted"). Instead it is cheaper to generate the list of completed tasks and query the list.
- Checks for cycles in graph

philipppahl committed:
Removed ExecutionGraph from DecisionInput; ExecutionGraph independent of tasks, only uses ids; Implementation needs testing

philipppahl committed:
unit and integration tests

amatthies commented:
+1

philipppahl committed:
adapted examples

philipppahl committed:
removed TODO tags

amatthies commented on floto/specs/serializer.py:
```diff
@@ -0,0 +1,36 @@
+    class_name = parts[-1]
+    return getattr(sys.modules[module_name], class_name)
+
+def copy_dict(dict_, filter_keys=[]):
```
default argument is mutable. Don't do this.

philipppahl committed:
removed mutable default argument
